## **Necessary imports**

In [59]:
import re
import numpy as np
import pandas as pd
from datetime import datetime
import matplotlib.pyplot as plt
%matplotlib inline

## **Mounting Drive**

In [60]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


## **Initializing parent folder**

In [61]:
%cd /content/drive/MyDrive/IPL_ChatBot

/content/drive/MyDrive/IPL_ChatBot


## **CSV file path**

In [62]:
df_path = "/content/drive/MyDrive/IPL_ChatBot/IPL_Matches_Gravitas_AI_Problem_Statement_Data.csv"

## **Reading & Viewing csv file**

In [63]:
df = pd.read_csv(df_path)
df.head()

id  season       city        date                        team1  \
0   1    2017  Hyderabad  2017-04-05          Sunrisers Hyderabad   
1   2    2017       Pune  2017-04-06               Mumbai Indians   
2   3    2017     Rajkot  2017-04-07                Gujarat Lions   
3   4    2017     Indore  2017-04-08       Rising Pune Supergiant   
4   5    2017  Bangalore  2017-04-08  Royal Challengers Bangalore   

                         team2                  toss_winner toss_decision  \
0  Royal Challengers Bangalore  Royal Challengers Bangalore         field   
1       Rising Pune Supergiant       Rising Pune Supergiant         field   
2        Kolkata Knight Riders        Kolkata Knight Riders         field   
3              Kings XI Punjab              Kings XI Punjab         field   
4             Delhi Daredevils  Royal Challengers Bangalore           bat   

   result  dl_applied                       winner  win_by_runs  \
0  normal           0          Sunrisers Hyderabad           35   
1  normal           0       Rising Pune Supergiant            0   
2  normal           0        Kolkata Knight Riders            0   
3  normal           0              Kings XI Punjab            0   
4  normal           0  Royal Challengers Bangalore           15   

   win_by_wickets player_of_match                                      venue  \
0               0    Yuvraj Singh  Rajiv Gandhi International Stadium, Uppal   
1               7       SPD Smith    Maharashtra Cricket Association Stadium   
2              10         CA Lynn     Saurashtra Cricket Association Stadium   
3               6      GJ Maxwell                     Holkar Cricket Stadium   
4               0       KM Jadhav                      M Chinnaswamy Stadium   

          umpire1        umpire2 umpire3  
0     AY Dandekar       NJ Llong     NaN  
1  A Nand Kishore         S Ravi     NaN  
2     Nitin Menon      CK Nandan     NaN  
3    AK Chaudhary  C Shamshuddin     NaN  
4             NaN            NaN     NaN

## **Shape**

In [64]:
df.shape

(756, 18)

## **Columns**

In [65]:
df.columns

Index(['id', 'season', 'city', 'date', 'team1', 'team2', 'toss_winner',
       'toss_decision', 'result', 'dl_applied', 'winner', 'win_by_runs',
       'win_by_wickets', 'player_of_match', 'venue', 'umpire1', 'umpire2',
       'umpire3'],
      dtype='object')

## **Checking if NaN value exists**

In [66]:
df.isna().sum()

id                   0
season               0
city                 7
date                 0
team1                0
team2                0
toss_winner          0
toss_decision        0
result               0
dl_applied           0
winner               4
win_by_runs          0
win_by_wickets       0
player_of_match      4
venue                0
umpire1              2
umpire2              2
umpire3            637
dtype: int64

### **We can see that there are 637 None Values in 'umpire3' column. These values can't be filled with other values. So, deleting this column.**

In [67]:
del df['umpire3']

## **Still NaN value exists in other columns**

In [68]:
df.isna().sum()

id                 0
season             0
city               7
date               0
team1              0
team2              0
toss_winner        0
toss_decision      0
result             0
dl_applied         0
winner             4
win_by_runs        0
win_by_wickets     0
player_of_match    4
venue              0
umpire1            2
umpire2            2
dtype: int64

## **Dropping all the NaN values**

In [69]:
df = df.dropna()

## **Shape after deleting**

In [70]:
df.shape

(743, 17)

## **Season, team1, team2 unique value counts**

In [71]:
df['season'].value_counts(), df['team1'].value_counts(), df['team2'].value_counts()

(season
 2013    76
 2012    74
 2011    72
 2010    60
 2016    60
 2018    60
 2017    58
 2008    58
 2019    58
 2009    57
 2015    57
 2014    53
 Name: count, dtype: int64,
 team1
 Mumbai Indians                 99
 Kings XI Punjab                91
 Chennai Super Kings            88
 Kolkata Knight Riders          82
 Royal Challengers Bangalore    81
 Delhi Daredevils               70
 Rajasthan Royals               67
 Sunrisers Hyderabad            60
 Deccan Chargers                43
 Pune Warriors                  20
 Gujarat Lions                  14
 Rising Pune Supergiant          8
 Kochi Tuskers Kerala            7
 Rising Pune Supergiants         7
 Delhi Capitals                  6
 Name: count, dtype: int64,
 team2
 Kolkata Knight Riders          95
 Royal Challengers Bangalore    93
 Delhi Daredevils               86
 Mumbai Indians                 85
 Kings XI Punjab                84
 Rajasthan Royals               77
 Chennai Super Kings            74
 Sunrise

## **City value counts**

In [72]:
df['city'].value_counts()

city
Mumbai            101
Kolkata            77
Delhi              73
Hyderabad          64
Bangalore          63
Chennai            57
Jaipur             47
Chandigarh         46
Pune               38
Durban             15
Bengaluru          13
Centurion          12
Ahmedabad          12
Visakhapatnam      12
Rajkot             10
Mohali             10
Indore              9
Dharamsala          9
Johannesburg        8
Cuttack             7
Ranchi              7
Port Elizabeth      7
Cape Town           7
Abu Dhabi           7
Sharjah             6
Raipur              6
Kochi               5
Kanpur              4
Nagpur              3
Kimberley           3
East London         3
Bloemfontein        2
Name: count, dtype: int64

## **Venue Counts**

In [73]:
df['venue'].value_counts()

venue
Eden Gardens                                            77
Wankhede Stadium                                        73
M Chinnaswamy Stadium                                   70
Feroz Shah Kotla                                        66
Rajiv Gandhi International Stadium, Uppal               56
MA Chidambaram Stadium, Chepauk                         49
Sawai Mansingh Stadium                                  47
Punjab Cricket Association Stadium, Mohali              35
Maharashtra Cricket Association Stadium                 21
Subrata Roy Sahara Stadium                              17
Dr DY Patil Sports Academy                              17
Kingsmead                                               15
Punjab Cricket Association IS Bindra Stadium, Mohali    14
Sardar Patel Stadium, Motera                            12
SuperSport Park                                         12
Brabourne Stadium                                       11
Dr. Y.S. Rajasekhara Reddy ACA-VDCA Cricket Stadiu

## **Toss decision and Result**

In [74]:
df['toss_decision'].value_counts(), df['result'].value_counts()

(toss_decision
 field    455
 bat      288
 Name: count, dtype: int64,
 result
 normal    734
 tie         9
 Name: count, dtype: int64)

## **date value counts**

In [75]:
df['date'].value_counts()

date
2013-04-06    2
2011-04-29    2
2011-05-01    2
2011-05-02    2
2011-05-04    2
             ..
2012-04-05    1
2012-04-04    1
2011-05-28    1
2011-05-27    1
12/05/19      1
Name: count, Length: 538, dtype: int64

## **There are some values in "day/month/year" format. But we want all the values in the same format "Year-Month-Day".**

In [76]:
date_pattern = re.compile(r"\b\d{2}/\d{2}/\d{2}\b")

for i in range(len(df)):
  date = df.iloc[i]['date']
  if date_pattern.match(date):
    date_obj = datetime.strptime(date, "%d/%m/%y")
    formatted_date = date_obj.strftime("%Y-%m-%d")
    df.iloc[i, df.columns.get_loc('date')] = formatted_date

## **Viewing the change**

In [77]:
df['date'].value_counts().to_dict()

{'2013-04-06': 2,
 '2011-04-29': 2,
 '2011-05-01': 2,
 '2011-05-02': 2,
 '2011-05-04': 2,
 '2011-05-05': 2,
 '2011-05-07': 2,
 '2011-05-08': 2,
 '2011-05-10': 2,
 '2016-05-08': 2,
 '2016-05-07': 2,
 '2011-05-14': 2,
 '2011-05-15': 2,
 '2016-05-01': 2,
 '2011-05-22': 2,
 '2016-04-30': 2,
 '2012-04-06': 2,
 '2012-04-07': 2,
 '2012-04-08': 2,
 '2016-04-24': 2,
 '2012-04-10': 2,
 '2016-04-23': 2,
 '2012-04-12': 2,
 '2012-04-19': 2,
 '2012-04-15': 2,
 '2012-04-17': 2,
 '2012-05-10': 2,
 '2012-04-21': 2,
 '2012-04-22': 2,
 '2016-04-17': 2,
 '2016-04-16': 2,
 '2011-04-30': 2,
 '2011-04-27': 2,
 '2012-04-29': 2,
 '2016-05-14': 2,
 '2018-04-22': 2,
 '2010-03-25': 2,
 '2018-04-21': 2,
 '2010-03-28': 2,
 '2010-03-31': 2,
 '2010-04-03': 2,
 '2010-04-04': 2,
 '2018-04-15': 2,
 '2018-04-14': 2,
 '2010-04-07': 2,
 '2010-04-10': 2,
 '2010-04-11': 2,
 '2010-04-13': 2,
 '2018-04-08': 2,
 '2010-04-17': 2,
 '2010-04-18': 2,
 '2016-05-22': 2,
 '2011-04-09': 2,
 '2011-04-10': 2,
 '2016-05-21': 2,
 '2011-04-

In [79]:
df.to_csv("IPL_chatbot_data.csv", index=False)